### ISSUE TITLE
autoencoder 

### ISSUE THREAD

**Issue by Chromer163** [MorvanZhou/tutorials#54]
我发现autoencoder代码里面的训练时候的权值一直没有改变。
很严重的错误

**Comment by MorvanZhou** [comment]
请问输出的 cost 有没有变化?

**Comment by Chromer163** [comment]
变化不大（基本不变），但是我输出了自己基于你的代码修改的的权值，发现不变。
然后去输出之前源码的权值，发现也不变。求解？
谢谢莫烦哥指导。

**Comment by Chromer163** [comment]
__#!/usr/bin/env python
# -*- coding: utf-8 -*-

import tensorflow as tf
import numpy as np
from ae_test import gain_linked_feature_list,  gain_label_list


# Parameters
learning_rate = 0.001
batch_size = 100
display_step = 100
examples_to_show = 10
training_epochs = 5000


# Network Parameters
n_hidden_1 = 300   # 1st layer num features
n_hidden_2 = 250    # 2nd layer num features
n_hidden_3 = 200    # 2nd layer num features
n_input = 3096     # input numbers
n_output = 3     # output numbers

# input data
feature_list = gain_linked_feature_list('/home/chromer/PycharmProjects/headpose/linked_feature')
train_feature_list = feature_list[:4800]
test_feature_list = feature_list[4800:]
label_list = gain_label_list("/home/chromer/PycharmProjects/headpose/label.txt")
train_label_list = label_list[:4800]
test_label_list = label_list[4800:]


# import data
def get_data(batch_size, batch_count, feature_list, label_list):
    feature_batch = feature_list[batch_count*batch_size:(batch_count+1)*batch_size]
    label_batch = label_list[batch_count*batch_size:(batch_count+1)*batch_size]
    batch_imgs = np.array(feature_batch)
    batch_labs = np.array(label_batch)
    return batch_imgs, batch_labs


# xavier initialize method
def xavier_init(fan_in, fan_out, constant=1):
    low = -constant * np.sqrt(6.0 / (fan_in + fan_out))
    high = constant * np.sqrt(6.0 / (fan_in + fan_out))
    return tf.random_uniform((fan_in, fan_out), minval=low, maxval=high, dtype=tf.float32)


weights = {
    'encoder_h1': tf.Variable(tf.random_normal([n_input, n_hidden_1]), name='w1'),
    'encoder_h2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]), name='w2'),
    'encoder_h3': tf.Variable(tf.random_normal([n_hidden_2, n_hidden_3]), name='w3'),
}
weights_h4 = tf.Variable(xavier_init(n_hidden_3, n_output), name='w4')

biases = {
    'encoder_b1': tf.Variable(tf.random_normal([n_hidden_1]), name='b1'),
    'encoder_b2': tf.Variable(tf.random_normal([n_hidden_2]), name='b2'),
    'encoder_b3': tf.Variable(tf.random_normal([n_hidden_3]), name='b3'),
}
biases_h4 = tf.Variable(tf.random_normal([n_output]), name='b4'),


# Building the encoder
def encoder(x):
    # Encoder Hidden layer with relu activation #1
    layer_1 = tf.nn.relu(tf.add(tf.matmul(x, weights['encoder_h1']),
                                   biases['encoder_b1']))
    # Encoder Hidden layer with relu activation #2
    layer_2 = tf.nn.relu(tf.add(tf.matmul(layer_1, weights['encoder_h2']),
                                   biases['encoder_b2']))
    # Encoder Hidden layer with relu activation #3
    layer_3 = tf.nn.relu(tf.add(tf.matmul(layer_2, weights['encoder_h3']),
                                    biases['encoder_b3']))
    # Encoder Hidden layer with full-connected activation #4
    layer_4 = tf.nn.sigmoid(tf.add(tf.matmul(layer_3, weights_h4), biases_h4))
    return layer_4

# Construct model
X_train = tf.placeholder(tf.float32, shape=(None, n_input))    # input images
Y_train = tf.placeholder(tf.float32, shape=(None, n_output))     # input labels
X_train = tf.div(tf.subtract(X_train, tf.reduce_min(X_train)),
                 tf.subtract(tf.reduce_max(X_train),  tf.reduce_min(X_train))
                 )
encoder_op_train = encoder(X_train)                  #output of encoder
# Prediction
y_pred_train = encoder_op_train           # predictions of encoders
y_true_train = Y_train                    # labels to calculate loss


# Define loss and optimizer, minimize the squared error
cost = tf.reduce_mean(tf.pow(y_true_train-y_pred_train, 2))        # cost between predictions and labels
optimizer = tf.train.AdamOptimizer(learning_rate).minimize(cost)


# Launch the graph
saver = tf.train.import_meta_graph('/home/chromer/PycharmProjects/headpose/pre_train1.ckpt-5000.meta')  # load model
with tf.Session() as sess:
    # Initializing the variables
    saver.restore(sess, tf.train.latest_checkpoint('/home/chromer/PycharmProjects/headpose'))    # load parameters
    sess.run(tf.global_variables_initializer())
    graph = tf.get_default_graph()
    # load pretrain parameters
    weights['encoder_h1'] = graph.get_tensor_by_name('w1:0')
    weights['encoder_h2'] = graph.get_tensor_by_name('w2:0')
    weights['encoder_h3'] = graph.get_tensor_by_name('w3:0')
    biases['encoder_b1'] = graph.get_tensor_by_name('b1:0')
    biases['encoder_b2'] = graph.get_tensor_by_name('b2:0')
    biases['encoder_b3'] = graph.get_tensor_by_name('b3:0')
    print('模型恢复成功')
    # numbers of batches of every epoch
    batch_number = int(len(train_label_list)/batch_size)
    # Training cycle
    batch_count = 0
    for epoch in range(training_epochs):
        train_feature, train_labels = get_data(batch_size, batch_count, train_feature_list, train_label_list)
        train_feature = np.reshape(train_feature, (batch_size, n_input))
        train_labels = np.reshape(train_labels, (batch_size, n_output))
        _, c, w = sess.run([optimizer, cost, weights_h4], feed_dict={X_train: train_feature, Y_train: train_labels})
        if batch_count < batch_number - 1:
            batch_count += 1
        else:
            batch_count = 0
        # Display logs per epoch step
        if epoch % display_step == 0:
            print("Epoch:", '%04d' % (epoch+1),
                  "cost=", "{:.9f}".format(c))
            print(w)
    print("Optimization Finished!")
    save_path = saver.save(sess, '/home/chromer/PycharmProjects/headpose/finetune.ckpt', training_epochs)
    print("Model saved in file: %s" % save_path)

**Comment by MorvanZhou** [comment]
这个 AE 的代码是会变的很慢, 如果你想看到快速变化的 cost, 建议提高 learning rate, (但是收敛不到很好的地方, 只是测试一下是不是真的在学).

**Comment by Chromer163** [comment]
关键是weights一点都不变，我的网络是根据你的AE代码改的，然后发现weights没更新，于是把你的github源码的weights也输出了一下，也没有更新，那么我想问这个网络是否有在训练？
这个是你的weights的输出结果：
[[ 0.71894276 -0.12779373 -0.47742644 ...,  0.3718285   1.94441688
  -0.80189955]
 [ 0.13756453  0.55470967 -0.58790702 ...,  0.85864913 -0.82666868
  -3.56348491]
 [ 1.33972335 -1.92569625  0.23700731 ...,  0.23490945 -0.97278339
  -0.07270477]
 ..., 
 [ 0.05120971 -1.46908712  0.18777686 ...,  1.11979592 -0.16183065
   0.42437086]
 [-0.52693421  1.04220462 -0.44850326 ..., -0.0170248   0.84664136
  -1.34703863]
 [ 0.38991982 -0.97013456  0.74403548 ...,  0.40549326  1.98958802
  -1.94241369]]
Epoch: 0002 cost= 0.082836024 [[ 0.71894276 -0.12779373 -0.47742644 ...,  0.3718285   1.94441688
  -0.80189955]
 [ 0.13756453  0.55470967 -0.58790702 ...,  0.85864913 -0.82666868
  -3.56348491]
 [ 1.33972335 -1.92569625  0.23700731 ...,  0.23490945 -0.97278339
  -0.07270477]
 ..., 
 [ 0.05120971 -1.46908712  0.18777686 ...,  1.11979592 -0.16183065
   0.42437086]
 [-0.52693421  1.04220462 -0.44850326 ..., -0.0170248   0.84664136
  -1.34703863]
 [ 0.38991982 -0.97013456  0.74403548 ...,  0.40549326  1.98958802
  -1.94241369]]
Epoch: 0003 cost= 0.078367628 [[ 0.71894276 -0.12779373 -0.47742644 ...,  0.3718285   1.94441688
  -0.80189955]
 [ 0.13756453  0.55470967 -0.58790702 ...,  0.85864913 -0.82666868
  -3.56348491]
 [ 1.33972335 -1.92569625  0.23700731 ...,  0.23490945 -0.97278339
  -0.07270477]
 ..., 
 [ 0.05120971 -1.46908712  0.18777686 ...,  1.11979592 -0.16183065
   0.42437086]
 [-0.52693421  1.04220462 -0.44850326 ..., -0.0170248   0.84664136
  -1.34703863]
 [ 0.38991982 -0.97013456  0.74403548 ...,  0.40549326  1.98958802
  -1.94241369]]
Epoch: 0004 cost= 0.071173191 [[ 0.71894276 -0.12779373 -0.47742644 ...,  0.3718285   1.94441688
  -0.80189955]
 [ 0.13756453  0.55470967 -0.58790702 ...,  0.85864913 -0.82666868
  -3.56348491]
 [ 1.33972335 -1.92569625  0.23700731 ...,  0.23490945 -0.97278339
  -0.07270477]
 ..., 
 [ 0.05120971 -1.46908712  0.18777686 ...,  1.11979592 -0.16183065
   0.42437086]
 [-0.52693421  1.04220462 -0.44850326 ..., -0.0170248   0.84664136
  -1.34703863]
 [ 0.38991982 -0.97013456  0.74403548 ...,  0.40549326  1.98958802
  -1.94241369]]
Epoch: 0005 cost= 0.069280788 [[ 0.71894276 -0.12779373 -0.47742644 ...,  0.3718285   1.94441688
  -0.80189955]
 [ 0.13756453  0.55470967 -0.58790702 ...,  0.85864913 -0.82666868
  -3.56348491]
 [ 1.33972335 -1.92569625  0.23700731 ...,  0.23490945 -0.97278339
  -0.07270477]
 ..., 
 [ 0.05120971 -1.46908712  0.18777686 ...,  1.11979592 -0.16183065
   0.42437086]
 [-0.52693421  1.04220462 -0.44850326 ..., -0.0170248   0.84664136
  -1.34703863]
 [ 0.38991982 -0.97013456  0.74403548 ...,  0.40549326  1.98958802
  -1.94241369]]

**Comment by MorvanZhou** [comment]
因为这份代码可能也受 tf 版本的影响, 我对2017年的 tf 做了一份新的教程. 如果你尝试我这一份新代码呢?  https://github.com/MorvanZhou/Tensorflow-Tutorial/blob/master/tutorial-contents/404_AutoEncoder.py

**Comment by Chromer163** [comment]
ok,我试试。

**Comment by daoyijushi** [comment]
请问我输入这个代码之后，在tensorboard上看不到任何的数据和图形是怎么回事

**Comment by MorvanZhou** [comment]
@daoyijushi  https://github.com/MorvanZhou/Tensorflow-Tutorial/blob/master/tutorial-contents/404_AutoEncoder.py  这个代码是没有 tensorboard 的, 需要自己加.
